sw/host/ot_certs/src/codegen.rs

```diff
@@ -0,0 +1,408 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module is capable of generating C code for generating a binary X.509
+//! certificate according to a [`Template`](crate::template::Template).
+
+use anyhow::{Context, Result};
+use heck::ToUpperCamelCase;
+use std::collections::HashMap;
+
+use crate::asn1::codegen::{self, ConstantPool, VariableCodegenInfo, VariableInfo};
+use crate::asn1::x509::X509;
+use crate::template::{EcdsaSignature, Signature, Template, Value, Variable, VariableType};
+
+const INDENT: &str = "  ";
+const WRITER_MAIN: &str = r#"
+#include <stdio.h>
+#include <stdlib.h>
+
+status_t fgetc_wrapper(void *context) {
+  int c = fgetc((FILE *)context);
+  if (c == EOF) return UNAVAILABLE(c);
+  else return OK_STATUS(c);
+}
+
+status_t putbuf_wrapper(void *context, const char *msg, size_t len) {
+  fwrite(msg, len, 1, stderr);
+  return OK_STATUS();
+}
+
+status_t execute_test(ujson_t uj, FILE *outf) {
+  values_t values;
+  TRY(ujson_deserialize_values_t(&uj, &values));
+  size_t cert_size = 1000;
+  uint8_t *cert = malloc(cert_size);
+  TRY(build_cert(&values, cert, &cert_size));
+  fwrite(cert, cert_size, 1, outf);
+  fclose(outf);
+  return OK_STATUS();
+}
+
+int main(int argc, char **argv) {
+  if (argc != 3) {
+    fprintf(stderr, "usage: %s <values.json> <output.pem>\n", argv[0]);
+    return 1;
+  }
+  FILE *inf = fopen(argv[1], "r");
+  if (inf == NULL) {
+    perror("cannot open json file with values");
+    return 2;
+  }
+  FILE *outf = fopen(argv[2], "wb");
+  if (outf == NULL) {
+    perror("cannot open output certificate file with values");
+    return 2;
+  }
+  ujson_t ujson = ujson_init(inf, fgetc_wrapper, putbuf_wrapper);
+  status_t err = execute_test(ujson, outf);
+  const char *code = NULL;
+  int32_t arg = 0;
+  char mod_id[5];
+  bool is_err = status_extract(err, &code, &arg, mod_id);
+  mod_id[4] = 0;
+  if(is_err) {
+    fprintf(stderr, "An error occured: %s[%d,%s]\n", code, arg, mod_id);
+  }
+  return 0;
+}
+"#;
```

Comment:
Nit / suggestion: you may find it clearer to put this in a `.c` file in this directory and use `include_str!` to pull it in.